rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -537,7 +537,74 @@ object Reduce {
                 .spatialMatch(substTarget, substPattern)
                 .runS(SpatialMatcher.emptyMap)
                 .isDefined))
-
+        case EPercentBody(EPercent(p1, p2)) =>
+          def evalToStringPair(keyExpr: Expr, valueExpr: Expr): M[(String, String)] =
+            (keyExpr.exprInstance, valueExpr.exprInstance) match {
+              case (GString(keyString), GString(valueString)) =>
+                Applicative[M].pure[(String, String)](keyString -> valueString)
+              case (GString(keyString), GInt(valueInt)) =>
+                Applicative[M].pure[(String, String)](keyString -> valueInt.toString)
+              // TODO: Add cases for other ground terms as well? Maybe it would be better
+              // to implement cats.Show for all ground terms.
+              case (_: GString, value) =>
+                s.raiseError[(String, String)](
+                  ReduceError(s"Error: interpolation doesn't support ${value.getClass}")
+                )
+              case _ =>
+                s.raiseError[(String, String)](
+                  ReduceError("Error: interpolation Map should only contain String keys")
+                )
+            }
+          for {
+            v1 <- evalSingleExpr(p1.get)
+            v2 <- evalSingleExpr(p2.get)
+            result <- (v1.exprInstance, v2.exprInstance) match {
+                       case (GString(lhs), EMapBody(ParMap(rhs, _, _))) =>
+                         rhs.iterator
+                           .map {
+                             case (k, v) =>
+                               for {
+                                 keyExpr   <- evalSingleExpr(k)
+                                 valueExpr <- evalSingleExpr(v)
+                                 result    <- evalToStringPair(keyExpr, valueExpr)
+                               } yield result
+                           }
+                           .toList
+                           .sequence[M, (String, String)]
+                           .map { keyValuePairs =>
+                             val result  = StringBuilder.newBuilder
+                             var current = lhs
+                             while (current.nonEmpty) {
+                               keyValuePairs.find {
+                                 case (k, v) => current.startsWith("${" + k + "}")
+                               } match {
+                                 case Some((k, v)) =>
+                                   result ++= v
+                                   current = current.drop(k.length + 3)
+                                 case None =>
+                                   result += current.head
+                                   current = current.tail
+                               }
+                             }
```

Comment:
Couldn't we just `replace` elements of format `${x}` in the string? 
https://scalafiddle.io/sf/nQz5nW1/0

Reply:
I also thought so (my second commit implementation looked just like yours), but unfortunately it is not so easy. Consider the following situation:
"Hello, ${name1}! Bye, ${name2}!" % {"name1": "Bob ${name2}", "name2": "Alice"}
Your implementation returns "Hello, Bob Alice! Bye, Alice!", while mine returns "Hello, Bob ${name2}! Bye, Alice!"

Reply:
Ok, makes sense. Can you interpolating part to the method?

Reply:
Also, I would prefer to see this implemented as a regex but it's just my personal taste. We can keep it as is.